### PR TITLE
Refactor lib/user out of analytics

### PR DESCRIPTION
**Pull request by jaswrks**
## Changes proposed in this Pull Request

First noted by @jsnajdr in https://github.com/Automattic/wp-calypso/pull/34832#discussion_r306340305

- Add `client/lib/analytics/utils/current-user.js`
- Refactor `lib/user` (and related `hashPii`, `getNormalizedHashedUserEmail`) out of analytics.
  - Instead, leverage `getCurrentUser()` in `client/lib/analytics/utils/current-user.js`.

### In an incognito window, open:
https://hash-5dc41bacf44763a27854707e9a4afd5c23c293d1.calypso.live/start/user

In DevTools console, run the following, then refresh the page:

```
document.cookie = 'flags=gdpr-banner,google-analytics,ad-tracking; path=/; domain=.' + document.location.hostname.split( '.' ).slice( -2 ).join( '.' );
document.cookie = 'sensitive_pixel_option=yes; path=/; domain=.' + document.location.hostname.split( '.' ).slice( -2 ).join( '.' );
localStorage.setItem( 'debug', 'calypso:*' );
```

- Confirm no unexpected JS console errors.
- Filter Network requests by `calypso_signup_start` and confirm 3 requests.
- Confirm `site_count=0`, `_ut=anon`, `_ui=[anon ID]` in Tracks pixel.

Screen Shot 2019-11-02 at 13 58 27

### Preserve Logs

- Set Network tab in DevTools to Preserve Log.
- Set Console tab in DevTools to Preserve Log.

---

### Complete the first step, Registration:

Screen Shot 2019-11-02 at 13 59 35

- Confirm no unexpected JS console errors.
- Filter Network requests by `calypso_user_registration_complete` and confirm 3 requests.
- Confirm `site_count=0`, `_ut=anon`, `_ui=[same anon ID]` in Tracks pixel.
  - Please be sure it's the **same anon ID** as in `calypso_signup_start`.

Screen Shot 2019-11-02 at 14 00 59

### Confirm Tracks `identifyUser()` working

Staying right where you are...

- Filter Network requests by `calypso_signup_actions_complete_step` and confirm 1 request.
- Confirm `site_count=0`, `_ut=wpcom:user_id`, `_ui=[WP.com numeric ID]` in Tracks pixel.

Screen Shot 2019-11-02 at 14 10 11

### Confirm Floodlight Aliasing

Staying right where you are...
This also confirms `setCurrentUser()`, `getCurrentUser()` are working.

- Filter Console tab by `floodlight`, and find `recordAliasInFloodlight`.
- Expand the data given and confirm `{ ID, username, email, hashedPii }` props exist.
- A little further down, look for `recordParamsInFloodlightGtag`. Expand data and confirm that the value of `u4` matches that of `hashedPii.ID` in aforementioned `recordAliasInFloodlight`.

Screen Shot 2019-11-02 at 14 04 53

---

### Complete Signup, stop at Checkout

Complete Signup by choosing a paid plan.
... but stop when you reach Checkout.

Screen Shot 2019-11-02 at 14 13 38

- Confirm no unexpected JS console errors.
- Filter network requests by `calypso_signup_complete` and confirm 3 requests.
- Confirm `site_count=1`, `_ut=wpcom:user_id`, `_ui=[WP.com numeric ID]` in Tracks pixel.
  - Also confirm `is_new_user: true`, `is_new_site: true` in Tracks pixel.

Screen Shot 2019-11-02 at 14 15 44

---

### Complete Checkout

Complete checkout now using store sandbox, in whichever way you prefer.

- Confirm no unexpected JS console errors.
- Filter network requests by `calypso_checkout_payment_success` and confirm 3 requests.
- Confirm `site_count=1`, `_ut=wpcom:user_id`, `_ui=[WP.com numeric ID]` in Tracks pixel.

Screen Shot 2019-11-02 at 14 27 20

---

### Test `PasswordlessSignupForm`

From `calypso.localhost:3000`, repeat, but this time put yourself into the passwordless A/B test group and confirm all the same events I noted previously.

Screen Shot 2019-11-02 at 14 33 39

---

@MicBosi We could use your eyes on all marketing analytics possibly impacted by this PR.

### CODE REVIEW

File: client/lib/analytics/utils/current-user.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+ * @returns {Ojbect|undefined} Current user.
+ */
+export function setCurrentUser( { ID, username, email } ) {
+	const userData = { ID: parseInt( ID, 10 ), username, email };
+
+	if ( ! userData.ID || isNaN( userData.ID ) || ! userData.username || ! userData.email ) {
+		return; // Invalid user data.
+	}
+
+	userData.hashedPii = {
```
Comment (jaswrks) [inline client/lib/analytics/utils/current-user.js]: @MicBosi Note this new approach to compiling hashed PII for marketing analytics.

Comment (jaswrks) [reply client/lib/analytics/utils/current-user.js]: Example use in `ad-tracking.js` looks like this.

```
const currentUser = getCurrentUser();
const params = currentUser ? { em: currentUser.hashedPii.email } : {};
```

File: client/lib/analytics/index.js [hunk 2]
```diff
@@ -180,16 +187,20 @@ if ( typeof window !== 'undefined' ) {
...
 
-		// this is called both for anonymous and logged-in users
-		checkForBlockedTracks();
-
+		// Identify current user.
 		if ( user && user.get() ) {
 			const userData = user.get();
-			analytics.identifyUser( userData.username, userData.ID );
+			initializeDebug( 'identifyUser', userData );
+			analytics.identifyUser( userData );
```
Comment (jsnajdr) [inline client/lib/analytics/index.js]: Let's do one more small patch on the `analytics.initialize` function: don't pass the instance of the `User` class, the one we need to call `.get()` on to get the user data, but pass the user data directly:
```js
analytics.initialize( currentUser.get(), ... );

function initialize( user ) {
  ...
  if ( user ) {
    analytics.identifyUser( user );
  }
  ...
}
```
That way, the Analytics library becomes even more independent from the Calypso's `lib/user` interfaces.

**Review by jsnajdr — APPROVED** [review]
Nice improvement, looks very good! :+1:

File: client/lib/analytics/index.js (hunk 2 above)
Comment (jaswrks) [reply client/lib/analytics/index.js]: Thanks. Updated in https://github.com/Automattic/wp-calypso/pull/37263/commits/5fb8ed78a6324c91d10c04692f4004c86b8c55d8

Comment (jaswrks) [reply client/lib/analytics/index.js]: @jsnajdr I appreciate you taking the time to carefully review this. Thanks again!

**Review by danielwrobert — COMMENTED** [review]
• It seems that I’m only getting 2 events in the network for any of the noted filters. Your test calls for 3 (https://cld.wthms.co/ddis5a).
• When running through in `calypso.localhost:3000` as `passwordless`, I get the same results up until “Confirm Floodlight Aliasing”. The `recordAliasInFloodlight` is set as an anonymous user so there is no ID set to match with the u4 property in recordParamsInFloodlightGtag. It also drops me into the dev docs so there’s no checkout steps. Are those the intended results? Seems like there should be an ID, looking at this line https://github.com/Automattic/wp-calypso/pull/37263/files#diff-fe395d984a496551eb0db8e55b862ddeR112.
• Seeing a few JS errors but that may not be related to this PR: I know this one is expected- https://cld.wthms.co/H5PA0U but not sure about this one- https://cld.wthms.co/lvqqlt.

**Comment by jaswrks** [comment]
@danielwrobert writes...
> It seems that I’m only getting 2 events in the network for any of the noted filters. Your test calls for 3 (https://cld.wthms.co/ddis5a).

In the screenshot it looks like Network requests are filtered down to IMG pixels only. Try setting this to "All" and you should find one more.

**Comment by jaswrks** [comment]
@danielwrobert This error is OK. On calypso.live the hash location has not been whitelisted is all. https://cld.wthms.co/H5PA0U

**Comment by jaswrks** [comment]
@danielwrobert  This error is more concerning, but I can't reproduce it. Was it a one-time thing or was it occurring consistently? https://cld.wthms.co/lvqqlt

**Comment by jaswrks** [comment]
@danielwrobert writes...
> When running through in calypso.localhost:3000 as passwordless, I get the same results up until “Confirm Floodlight Aliasing”. The recordAliasInFloodlight is set as an anonymous user so there is no ID set to match with the u4 property in recordParamsInFloodlightGtag. It also drops me into the dev docs so there’s no checkout steps.

This is looking OK on my side. Once you register (email address only in passwordless mode), then you shouldn't be anonymous any longer. There should be an ID in the console.log output for recordAliasInFloodlight, like this:
https://drop.jas.xyz/ICQrXV

Regarding dev docs. Did you start from: http://calypso.localhost:3000/start/user ?

**Review by MicBosi — APPROVED** [review]

**Review by danielwrobert — APPROVED** [review]
@jaswrks Went through both flows once more noting your points above.

Everything looks good re this PR! 🚢 

There are a couple things to note:

- I did not see that 502 error I noted above (https://cld.wthms.co/lvqqlt). I preserved the log and here are all of the errors that came up (which all seem expected): https://cld.wthms.co/wEKRBu.
- One thing that tripped me up during the `PasswordlessSignupForm` test was that a couple of the events seemed to be named differently from the normal signup flow:
  - The `calypso_user_registration_complete` event is named `calypso_user_registration_passwordless_complete`.
  - The `calypso_checkout_payment_success` seems to be showing as `calypso_cart_unrecognized_payment_method` - which _could_ be a result of me paying via credits that I set in the Store Admin? But I did this for both flow tests and it appeared as expected (`
calypso_checkout_payment_success`) in the normal signup flow.

**Comment by jaswrks** [comment]
cc @lsinger on this one, it contains that sales widget error.
https://cld.wthms.co/wEKRBu

**Pull request merged**